compiler/grainlsp/rpc.re

```diff
@@ -0,0 +1,424 @@
+  line: int,
+  character: int,
+};
+
+[@deriving yojson]
+type lens_t = {
+  line: int,
+  signature: string,
+};
+
+[@deriving yojson]
+type range_t = {
+  start_line: int,
+  start_char: int,
+  end_line: int,
+  end_char: int,
+};
+
+[@deriving yojson]
+type range = {
+  start: position,
+  [@key "end"]
+  range_end: position,
+};
+
+[@deriving yojson]
+type capabilities_result = {capabilities: lsp_capabilities};
+[@deriving yojson]
+type capabilities_response = {
+  jsonrpc: string,
+  id: int,
+  result: capabilities_result,
+};
+
+[@deriving yojson]
+type command_t = {
+  title: string,
+  command: string,
+};
+
+[@deriving yojson]
+type lsp_lens_t = {
+  range,
+  command: command_t,
+};
+
+[@deriving yojson]
+type lens_response = {
+  jsonrpc: string,
+  id: int,
+  result: list(lsp_lens_t),
+};
+
+[@deriving yojson]
+type diagnostic_t = {
+  range,
+  severity: int,
+  message: string,
+};
+
+[@deriving yojson]
+type document_diagnostics = {
+  uri: string,
+  diagnostics: list(diagnostic_t),
+};
+
+[@deriving yojson]
+type diagnostics_message = {
+  jsonrpc: string,
+  method: string,
+  params: document_diagnostics,
+};
+
+[@deriving yojson]
+type markup_content = {
+  kind: string,
+  value: string,
+};
+
+[@deriving yojson]
+type marked_string = {
+  language: string,
+  value: string,
+};
+
+[@deriving yojson]
+type hover_result = {
+  contents: markup_content,
+  range,
+};
+
+[@deriving yojson]
+type hover_response = {
+  jsonrpc: string,
+  id: int,
+  result: hover_result,
+};
+
+[@deriving yojson]
+type definition_result = {
+  uri: string,
+  range,
+};
+
+[@deriving yojson]
+type definition_response = {
+  jsonrpc: string,
+  id: int,
+  result: definition_result,
+};
+
+[@deriving yojson]
+type completion_item = {
+  label: string,
+  kind: int,
+  detail: string,
+  documentation: string,
+};
+
+[@deriving yojson]
+type completion_result = {
+  isIncomplete: bool,
+  items: list(completion_item),
+};
+
+[@deriving yojson]
+type completion_response = {
+  jsonrpc: string,
+  id: int,
+  result: completion_result,
+};
+
+let convert_range = range => {
+  let r_start: position = {
+    line: range.start_line - 1,
+    character: range.start_char,
+  };
+  let r_end: position = {line: range.end_line - 1, character: range.end_char};
+  {start: r_start, range_end: r_end};
+};
+
+let read_message = (input): protocol_msg => {
+  let clength = input_line(input);
+  let cl = "Content-Length: ";
+  let cll = String.length(cl);
+  if (String.sub(clength, 0, cll) == cl) {
+    /* if on windows, dont need the extra -1 */
+    let offset = Sys.os_type == "Win32" ? 0 : (-1); /* -1 for trailing \r */
+
+    let num =
+      String.sub(clength, cll, String.length(clength) - cll + offset);
+    let num = (num |> int_of_string) + (Sys.os_type == "Win32" ? 1 : 2);
+    let buffer = Buffer.create(num);
+    Buffer.add_channel(buffer, input, num);
+    let raw = Buffer.contents(buffer);
+    let json = Yojson.Safe.from_string(raw);
```

Comment:
There is, but because the protocol injects Content-Length headers into the stream, we have to read from the channel the number of characters specified.